distsys/resources/nestedarch.go

```diff
@@ -0,0 +1,344 @@
+			select {
+			case <-res.cachedTimer.C:
+			default:
+			}
+		}
+		res.cachedTimer.Reset(d)
+	}
+	return res.cachedTimer.C
+}
+
+func (res *nestedArchetype) performRequest(reqTpe tla.TLAValue, fields... tla.TLARecordField) (tla.TLAValue, error) {
+	select {
+	case res.sendCh <-tla.MakeTLARecord(append(fields, tla.TLARecordField{
+		Key: nestedArchetypeConstants.tpe,
+		Value: reqTpe,
+	})):
+		// go to next select
+	case err := <-res.ctxErrCh:
+		return tla.TLAValue{}, fmt.Errorf("encountered nested archetype error: %w", err)
+	}
+
+	select {
+	case resp := <-res.receiveCh:
+		return resp, nil
+	case err := <-res.ctxErrCh:
+		return tla.TLAValue{}, fmt.Errorf("encountered nested archetype error: %w", err)
+	}
+}
+
+func (res *nestedArchetype) catchTLATypeErrors(err *error) {
+	if p := recover(); p != nil {
+		if e, ok := p.(error); ok {
+			*err = fmt.Errorf("%w: %v", ErrNestedArchetypeProtocol, e)
+		} else {
+			panic(p)
+		}
+	}
+}
+
+func (res *nestedArchetype) performRequestOrAbort(reqTpe tla.TLAValue, fields... tla.TLARecordField) (tla.TLAValue, error) {
+	select {
+	case res.sendCh <- tla.MakeTLARecord(append(fields, tla.TLARecordField{
+		Key: nestedArchetypeConstants.tpe,
+		Value: reqTpe,
+	})):
+		// go to next select
+	case err := <-res.ctxErrCh:
+		return tla.TLAValue{}, fmt.Errorf("encountered nested archetype error: %w", err)
+	case <-res.ensureTimer(nestedArchetypeTimeout):
+		return tla.TLAValue{}, distsys.ErrCriticalSectionAborted
+	}
+
+	select {
+	case resp := <-res.receiveCh:
+		return resp, nil
+	case err := <-res.ctxErrCh:
+		return tla.TLAValue{}, fmt.Errorf("encountered nested archetype error: %w", err)
+	case <-res.ensureTimer(nestedArchetypeTimeout):
+		return tla.TLAValue{}, distsys.ErrCriticalSectionAborted
+	}
+}
+
+func (res *nestedArchetype) handleResponseValue(value tla.TLAValue, allowAborted bool, expectedTpes... tla.TLAValue) (err error){
+	defer res.catchTLATypeErrors(&err)
+
+	tpe := value.ApplyFunction(nestedArchetypeConstants.tpe)
+	// map "aborted" over to a critical section abort. we'll still stop what we're doing, but we'll report the
+	// non-fatal error, which, according to allowAborted, is what we mean.
+	if allowAborted && tpe.Equal(nestedArchetypeAborted) {
+		return distsys.ErrCriticalSectionAborted
+	}
+
+	for _, expectedTpe := range expectedTpes {
+		if tpe.Equal(expectedTpe) {
+			return nil
+		}
+	}
+	return fmt.Errorf("%w: archetype response had unexpected tag %v, expected %v", ErrNestedArchetypeProtocol, tpe, expectedTpes)
+}
+
+func (res *nestedArchetype) Abort() chan struct{} {
+	res.assertSanity(false)
+	go func() {
+		staleAcksReceived := 0
+
+		retryReq:
+		select {
+		case res.sendCh <-tla.MakeTLARecord([]tla.TLARecordField{
+			{nestedArchetypeConstants.tpe, nestedArchetypeAbortReq },
+		}):
+			// go to next select, we successfully sent the abort request
+		case resp := <-res.receiveCh:
```

Comment:
I should add a reference to the `assertSanity` doc comment here / add more docs. I think there used to be a better comment for this situation, but it was probably removed between the many versions of this piece of code.

This case is because, uniquely, `Abort` might be sent before the nested archetype(s) finish processing a request, due to a timeout waiting for a response. In that case, we should read and discard the response without complaining (then go back and try again to send our abort request, because select semantics, as the request channel is not buffered), then wait as long as we need for the abort to be serviced.

Reply:
I see, it makes sense. Adding comments definitely helps.